### PR TITLE
extract code from text; solve_problem; request_timeout in config; improve code

### PR DESCRIPTION
**Pull request by sonichi**
## Why are these changes needed?

* added utility functions to extract code from text and generate code.
* added utility function to solve math problem.
* handle request_timeout in config.
* added `improve_function`, `improve_code` and refactored some existing AI functions.
* added test and documentation about utility functions.

## Checks

- [x] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR, or I've made sure [lint with flake8](https://github.com/microsoft/FLAML/blob/816a82a1155b4de4705b21a615ccdff67c6da379/.github/workflows/python-package.yml#L54-L59) output is two 0s.
- [x] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: test/openai/test_completion.py [hunk 1]
```diff
@@ -235,5 +273,5 @@ def my_average(results):
 
     openai.api_key_path = "test/openai/key.txt"
     test_nocontext()
-    test_humaneval(1)
-    test_math(1)
+    # test_humaneval(1)
```
Comment (qingyun-wu) [inline test/openai/test_completion.py:276]: Why comment these two tests?

File: flaml/version.py [hunk 2]
```diff
@@ -1 +1 @@
-__version__ = "1.2.1"
+__version__ = "1.2.2"
```
Comment (qingyun-wu) [inline flaml/version.py:1]: Do we need to update the version for this change?

File: test/openai/test_completion.py (hunk 1 above)
Comment (sonichi) [reply test/openai/test_completion.py:276]: These don't matter because pytest will test all of them.

File: flaml/version.py (hunk 2 above)
Comment (sonichi) [reply flaml/version.py:1]: Yes, to distinguish from the released version. It doesn't mean we'll release right after merging.

Comment (sonichi) [reply flaml/version.py:1]: With the new changes, I'd like to release a new version after merging.

**Review by qingyun-wu — APPROVED** [review]
Looks good to me.

**Comment by skzhang1** [comment]
I do not find problems/potential errors in the current version.

**Review by skzhang1 — APPROVED** [review]

File: flaml/autogen/code_utils.py [hunk 3]
```diff
@@ -1,56 +1,260 @@
...
 import sys
+import os
 from typing import List, Dict, Tuple, Optional, Union, Callable
-from flaml import oai
+import re
+import time
+from flaml.autogen import oai, DEFAULT_MODEL, FAST_MODEL
+
+# Regular expression for finding a code block
+CODE_BLOCK_PATTERN = r"```python\n(.*?)\n```"
```
Comment (BeibinLi) [inline flaml/autogen/code_utils.py:11]: Is the code always in python? What if the markdown doesn't contain the keyword "python" in the code block? 

How about change "python" to "\w*"? 

r"```\w*\n(.*?)\n```"

**Review by BeibinLi — COMMENTED** [review]
LGTM

**Review by BeibinLi — APPROVED** [review]
Looks good to me~

**Pull request merged**